ReadMe
                       Rosetta-Engine-Generic
----------------------------------------------------------------------

by Darren Duncan <perl@DarrenDuncan.net>

CONTENTS

* Preface
* Keeping Up To Date
* Dependencies
* Support
* License And Copyright
* Acknowledgements

PREFACE

This distribution features the Perl 5 module named
"Rosetta::Engine::Generic"; see the file lib/Rosetta/Engine/Generic.pm for
the main documentation, which is in POD format.  The distribution also
includes a ChangeLog file to say what changes already happened, and a TODO
file to say what changes have yet to happen.  The file INSTALL should help
with installation issues.  Read below for how to get support or keep up to
date, know this distribution's dependencies, or read the credits or
copyright and license.

This distribution requires a live database to be tested against.  To gain
the best understanding of whether Rosetta::Engine::Generic will work with
your preferred database engine and/or working environment, you should edit
this distribution's test configuration file "t_setup.pl" prior to running
'make test' and/or 'perl Makefile.PL'.  The setup details that you input
should match a visible database engine that you have full privileges on,
including the ability to create schema objects, and select or modify data.
If you do not edit "t_setup.pl", such as when you are a fully-automatic
CPAN tester, then the test suite will use the newest SQLite v3 database
engine by default, and the executed tests will only be as thorough as the
SQLite v3 feature set supports.

KEEPING UP TO DATE

My module set is constantly under development.  The canonical copies are
all linked to through my website, "http://www.DarrenDuncan.net/", on the
page called "Perl Libraries I Made" (name subject to change).

For those of you that are interested in the bleeding edge, I now have a
public Subversion repository at "http://svn.utsl.gen.nz/trunk/Rosetta/";
this repository contains all of my newer modules (whose problem domain is
databases), but not my older modules (which are not about databases).

You can download slightly older but more stable versions of all of my
modules in standard tarred and gzipped distributions, either from the
public CPAN archives, "http://www.cpan.org/authors/id/D/DU/DUNCAND/", or
from my own web server, "http://www.DarrenDuncan.net/d/perl/".

CPAN also extracts the POD from modules and maintains that documentation in
an organized and searchable fashion.  Their main search site is
"http://search.cpan.org/", and my own DUNCAND author page is
"http://search.cpan.org/~duncand/"; distributions are also there.

You can download my oldest distributions from BACKPAN, at
"http://history.perl.org/backpan/authors/id/D/DU/DUNCAND/", or from my web
server, at "http://www.DarrenDuncan.net/d/perl/archives/".

DEPENDENCIES

This distribution requires any version of Perl 5.x.y that is at least
5.8.1.

It also requires the Perl modules 'version' and 'only', which would
conceptually be built-in to Perl, but aren't, so they are on CPAN instead.

It also requires the Perl module List::Util, which would conceptually be
built-in to Perl, but is bundled with it instead.

It also requires these modules that are on CPAN: Rosetta '0.48.0-',
Rosetta::Validator '0.48.0-' (but only for the test suite),
SQL::Routine::SQLBuilder '0.21.0-', SQL::Routine::SQLParser '0.2.0-', DBI
'1.48-', DBD::SQLite '1.09-' (default database engine used by the test
suite), various DBI Driver (DBD) modules.

This distribution also requires Test::More for its test suite, and that
module is bundled with Perl.

SUPPORT

Currently I don't have any support arranged with other people, lists,
newsgroups, or otherwise.  Feel free to ask me if you can't figure things
out on your own, or another person whom you know has used this.  I may
start a mailing list for support issues later, so that users of my modules
can help each other with them.

LICENCE AND COPYRIGHT

This file is part of the Rosetta::Engine::Generic feature reference
implementation of the Rosetta database portability library.

Rosetta::Engine::Generic is Copyright (c) 2002-2005, Darren R. Duncan.  All
rights reserved.  Address comments, suggestions, and bug reports to
perl@DarrenDuncan.net, or visit http://www.DarrenDuncan.net/ for more
information.

Rosetta::Engine::Generic is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License (GPL) as
published by the Free Software Foundation (http://www.fsf.org/); either
version 2 of the License, or (at your option) any later version.  You
should have received a copy of the GPL as part of the
Rosetta::Engine::Generic distribution, in the file named "GPL"; if not,
write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
Boston, MA  02110-1301, USA.

Linking Rosetta::Engine::Generic statically or dynamically with other
modules is making a combined work based on Rosetta::Engine::Generic.  Thus,
the terms and conditions of the GPL cover the whole combination.  As a
special exception, the copyright holders of Rosetta::Engine::Generic give
you permission to link Rosetta::Engine::Generic with independent modules,
regardless of the license terms of these independent modules, and to copy
and distribute the resulting combined work under terms of your choice,
provided that every copy of the combined work is accompanied by a complete
copy of the source code of Rosetta::Engine::Generic (the version of
Rosetta::Engine::Generic used to produce the combined work), being
distributed under the terms of the GPL plus this exception.  An independent
module is a module which is not derived from or based on
Rosetta::Engine::Generic, and which is fully useable when not linked to
Rosetta::Engine::Generic in any form.

Any versions of Rosetta::Engine::Generic that you modify and distribute
must carry prominent notices stating that you changed the files and the
date of any changes, in addition to preserving this original copyright
notice and other credits.  Rosetta::Engine::Generic is distributed in the
hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

While it is by no means required, the copyright holders of
Rosetta::Engine::Generic would appreciate being informed any time you
create a modified version of Rosetta::Engine::Generic that you are willing
to distribute, because that is a practical way of suggesting improvements
to the standard version.

ACKNOWLEDGEMENTS

None besides myself so far.

FIN

Share and Enjoy!


Changes
                       Rosetta-Engine-Generic
----------------------------------------------------------------------

2005-09-28   Darren Duncan <perl@DarrenDuncan.net>

    * Rosetta-Engine-Generic Release 0.21.2 (SVN r878).

    * New code file versions are: Generic.pm 0.21.2 and en.pm 0.13.2.

    * From now on, any ordinary 'use [|only ]Foo' statements that appear
    within *.pm files will be located in different places than before; the
    'use' statements for modules that do not export anything, particularly
    object oriented modules, will appear near the top of the file, above
    all package declarations, but just below the use-pragma statements; the
    'use' statements for modules whose exported functions we are using will
    be placed just below the declarations of each package in which the
    relevant exported functions are used.

    * Removed any line-trailing whitespace from all distribution files.

    * Reformatted all code by swapping various string quoting delimiters.

    * Reformatted Generic.pm and Rosetta_Engine_Generic.t to un-cuddle all
    'else' and 'elsif' blocks.

    * Updated Generic.pm and Rosetta_Engine_Generic.t to replace all of
    their 'unless' conditionals with equivalent 'if' conditionals.

    * Updated Generic.pm and Rosetta_Engine_Generic.t to rename all of
    their 'foreach' loops to 'for'.

    * Updated Rosetta_Engine_Generic.t to move its one switch statement
    label upwards to its own line.

    * Updated Generic.pm to reformat the 9 '|' delimited regular
    expressions as '/' delimited ones.

    * Updated Generic.pm to reformat all 9 regular expressions so that
    they use the /x flag.

    * Reformatted all code to change any "if( ... ) {" to "if (...) {".

    * Added named constant $EMPTY_STR to Generic.pm.

    * Updated Generic.pm to add labels to all un-labeled [next|last|redo]
    statements (5), and their innermost bounding loop controls (3); the
    conditional of another un-labeled 'next' statement was merged into a
    subsequent 'if' instead.

    * Reformatted all code so that any uses of the named unary operators
    [defined|ref] no longer have parenthesis around their argument.

    * Updated Generic.pm to reformat 1 use of 'uc', 2 uses of 'lc', and 1
    use of 'ucfirst', like with 'ref'.

    * Removed 1 superfluous use of 'scalar' in Rosetta_Engine_Generic.t.

    * In Generic.pm, split up 6 multi-statement lines so each statement is
    on its own line.

    * Updated Generic.pm to replace 7 "if ... return ..." with their
    postfix-if equivalents; Rosetta_Engine_Generic.t had 1 such change.

    * Updated Generic.pm and Rosetta_Engine_Generic.t to rearrange (1,1)
    conditional 'die' statements so they have postfix-if conditionals.

    * Updated Generic.pm to rearrange 2 conditional '_throw...' statements
    so they have postfix-if conditionals.

    * Reformatted all code, in Generic.pm and Rosetta_Engine_Generic.t, so
    that any uses of the built-in operators or functions named
    [bless|join|push|shift|split|unpack] no longer have parenthesis around
    their argument lists; either the parens were just removed, or they were
    moved to surround both the operator/function name and its arguments.

    * Updated the install_dbi_driver() method of Generic.pm to merge two
    for-loops such that each available driver name is tried that matches
    the driver hint pattern until one succeeds in loading; previously, if
    the first driver matched by the pattern failed to load, the function
    didn't try any others.

    * Added new external dependency on the honorary built-in function
    List::Util::first() to Generic.pm.  Then replaced a last-terminated
    for-loop in srtn_catalog_list() with an invocation of it.

    * Updated any DEPENDENCIES documentation to reformat any indented
    sections as paragraphs, and show version ranges like 'only' takes them.

    * Reformatted all code so that every occurance of the string
    concatenation operator (.) has a space between the operator and each of
    its 2 arguments, rather than their all being in contact.  Also
    reformatted any lines that are split on this operator so the operator
    appears at the start of the second line, rather than the end of the
    first line.

    * Reformatted a few multi-line statements so that any [=|and|or] that
    are split on appear at the start of a line rather than the end of one.

    * Reformatted all of the code comments at the top of Generic.pm,
    where its object property names, and constant values, are declared,
    so that all comment lines don't exceed the 75 character line length.

    * In Generic.pm, reformatted 1 cascading ternary statement into
    aligned columns.

    * Other miscellaneous code line alignments and splitting.

2005-09-12   Darren Duncan <perl@DarrenDuncan.net>

    * Rosetta-Engine-Generic Release 0.21.1 (SVN r808).

    * New code file versions are: Generic.pm 0.21.1 and en.pm 0.13.1.

    * Updated all POD-containing files to re-wrap any non-indented POD
    paragraphs to a 75 character width, which is 5 less than the 80
    character width they were wrapped to before.  This change should make
    it easier to copy and paste a diff or patch of this documentation into
    an email message, where it is quoted at least once, without any line
    wrapping occurring.  Likewise, these standard documentation files were
    re-wrapped to 75 characters: ReadMe, INSTALL, LGPL|GPL, Changes, TODO.

    * Revised this Changes file to replace all detail entries for releases
    0.18 thru 0.21.0 with a significant release list for the same period;
    also added references to when all 'Rosetta developer release' (#s 1-3)
    were.

2005-09-08   Darren Duncan <perl@DarrenDuncan.net>

    * Rosetta-Engine-Generic Release 0.21.0 (SVN r787), containing
    Rosetta::Engine::Generic 0.21.0, was released on CPAN.  This is the
    first release of this distribution and its modules for which they had

lib/Rosetta/Engine/Generic/L/en.pm
#!perl
use 5.008001; use utf8; use strict; use warnings;

package Rosetta::Engine::Generic::L::en;
use version; our $VERSION = qv('0.13.2');

######################################################################

my $CG = 'Rosetta::Engine::Generic';

my %text_strings = (
    'ROS_G_NO_DBI_DRIVER_HINT_MATCH' =>
        $CG . q[ - can't find any installed DBI driver with a name like "{NAME}"],

    'ROS_G_RAW_SQLSTATE' =>
        $CG . q[ - following the most recent externally invocated procedure, the returned ]
        . q["SQLSTATE" 5-character-string value was "{SQLSTATE_NUM}" (zero is success)],
);

######################################################################

sub get_text_by_key {
    my (undef, $msg_key) = @_;
    return $text_strings{$msg_key};
}

######################################################################

1;
__END__

=encoding utf8

=head1 NAME

Rosetta::Engine::Generic::L::en - Localization of Rosetta::Engine::Generic for English

=head1 VERSION

This document describes Rosetta::Engine::Generic::L::en version 0.13.2.

=head1 SYNOPSIS

    use Locale::KeyedText;
    use Rosetta;

    # do work ...

    my $translator = Locale::KeyedText->new_translator(
        ['Rosetta::Engine::Generic::L::', 'SQL::Routine::SQLBuilder::L::',
        'SQL::Routine::SQLParser::L::', 'Rosetta::L::', 'SQL::Routine::L::'], ['en'] );

    # do work ...

    eval {
        # do work with Rosetta, which may throw an exception ...
    };
    if (my $error_message_object = $@) {
        # examine object here if you want and programmatically recover...

        # or otherwise do the next few lines...
        my $error_user_text = $translator->translate_message( $error_message_object );
        # display $error_user_text to user by some appropriate means
    }

    # continue working, which may involve using Rosetta some more ...

=head1 DESCRIPTION

The Rosetta::Engine::Generic::L::en Perl 5 module contains localization
data for the Rosetta::Engine::Generic module.  It complements the
Rosetta::L::en module, which should interpret most messages that Generic
throws; Generic::L just contains extra or overridden messages specific to
Generic.  It is designed to be interpreted by Locale::KeyedText.

This class is optional and you can still use Rosetta::Engine::Generic
effectively without it, especially if you plan to either show users
different error messages than this class defines, or not show them anything
because you are "handling it".

=head1 FUNCTIONS

=head2 get_text_by_key( MSG_KEY )

    my $user_text_template = Rosetta::Engine::Generic::L::en->get_text_by_key( 'foo' );

This function takes a Message Key string in MSG_KEY and returns the
associated user text template string, if there is one, or undef if not.

=head1 DEPENDENCIES

This module requires any version of Perl 5.x.y that is at least 5.8.1.

It also requires the Perl module L<version>, which would conceptually be
built-in to Perl, but isn't, so it is on CPAN instead.

This module has no enforced dependencies on L<Locale::KeyedText>, which is
on CPAN, or on L<Rosetta::Engine::Generic>, which is in the current
distribution, but it is designed to be used in conjunction with them.

=head1 INCOMPATIBILITIES

None reported.

=head1 SEE ALSO

L<perl(1)>, L<Locale::KeyedText>, L<Rosetta::Engine::Generic>.

=head1 BUGS AND LIMITATIONS

The structure of this module is trivially simple and has no known bugs.

However, the locale data that this module contains may be subject to large
changes in the future; you can determine the likeliness of this by
examining the development status and/or BUGS AND LIMITATIONS documentation
of the other module that this one is localizing; there tends to be a high
correlation in the rate of change between that module and this one.

=head1 AUTHOR

Darren R. Duncan (C<perl@DarrenDuncan.net>)

=head1 LICENCE AND COPYRIGHT

This file is part of the Rosetta::Engine::Generic feature reference
implementation of the Rosetta database portability library.

Rosetta::Engine::Generic is Copyright (c) 2002-2005, Darren R. Duncan.  All
rights reserved.  Address comments, suggestions, and bug reports to
C<perl@DarrenDuncan.net>, or visit L<http://www.DarrenDuncan.net/> for more
information.

Rosetta::Engine::Generic is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License (GPL) as
published by the Free Software Foundation (L<http://www.fsf.org/>); either
version 2 of the License, or (at your option) any later version.  You
should have received a copy of the GPL as part of the
Rosetta::Engine::Generic distribution, in the file named "GPL"; if not,
write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
Boston, MA  02110-1301, USA.

Linking Rosetta::Engine::Generic statically or dynamically with other
modules is making a combined work based on Rosetta::Engine::Generic.  Thus,
the terms and conditions of the GPL cover the whole combination.  As a
special exception, the copyright holders of Rosetta::Engine::Generic give
you permission to link Rosetta::Engine::Generic with independent modules,
regardless of the license terms of these independent modules, and to copy
and distribute the resulting combined work under terms of your choice,
provided that every copy of the combined work is accompanied by a complete
copy of the source code of Rosetta::Engine::Generic (the version of
Rosetta::Engine::Generic used to produce the combined work), being
distributed under the terms of the GPL plus this exception.  An independent
module is a module which is not derived from or based on
Rosetta::Engine::Generic, and which is fully useable when not linked to
Rosetta::Engine::Generic in any form.

Any versions of Rosetta::Engine::Generic that you modify and distribute
must carry prominent notices stating that you changed the files and the
date of any changes, in addition to preserving this original copyright
notice and other credits.  Rosetta::Engine::Generic is distributed in the
hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

While it is by no means required, the copyright holders of
Rosetta::Engine::Generic would appreciate being informed any time you
create a modified version of Rosetta::Engine::Generic that you are willing
to distribute, because that is a practical way of suggesting improvements
to the standard version.

=cut


Makefile.PL
#!perl
use 5.008001; use utf8; use strict; use warnings;

use ExtUtils::MakeMaker;

warn <<__endquote;
------------------------------------------------------------
This distribution requires a live database to be tested against.  To gain the
best understanding of whether Rosetta::Engine::Generic will work with your
preferred database engine and/or working environment, you should edit this
distribution's test configuration file "t_setup.pl" prior to running 'make test'
and/or 'perl Makefile.PL'.  The setup details that you input should match a
visible database engine that you have full privileges on, including the ability
to create schema objects, and select or modify data.  If you do not edit
"t_setup.pl", such as when you are a fully-automatic CPAN tester, then the test
suite will use the newest SQLite v3 database engine by default, and the executed
tests will only be as thorough as the SQLite v3 feature set supports.
------------------------------------------------------------
__endquote

WriteMakefile(
    'NAME' => 'Rosetta::Engine::Generic',
    'VERSION_FROM' => 'lib/Rosetta/Engine/Generic.pm', # finds $VERSION
    'PREREQ_PM' => {
        'version' => 0,
        'only' => 0,
        'List::Util' => 0,
        'Test::More' => 0, # but only for the test suite
        'Rosetta' => '0.48.0',
        'Rosetta::Validator' => '0.48.0', # but only for the test suite
        'SQL::Routine::SQLBuilder' => '0.21.0',
        'SQL::Routine::SQLParser' => '0.2.0',
        'DBI' => '1.48',
        'DBD::SQLite' => '1.09', # default database engine used by the test suite
    },
);

1;
